File: jadx_out/sources/com/zhufucdev/motion_emulator/data/MotionRecorder.java
package com.zhufucdev.motion_emulator.data;

import android.content.Context;
import android.hardware.Sensor;
import android.hardware.SensorEvent;
import android.hardware.SensorEventListener;
import android.hardware.SensorManager;
import com.aventrix.jnanoid.jnanoid.NanoIdUtils;
import com.zhufucdev.me.stub.Motion;
import com.zhufucdev.me.stub.MotionMoment;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;
import kotlin.Metadata;
import kotlin.Unit;
import kotlin.collections.CollectionsKt;
import kotlin.jvm.functions.Function1;
import kotlin.jvm.internal.Intrinsics;
import kotlin.jvm.internal.Ref;

/* JADX INFO: compiled from: MotionRecorder.kt */
/* JADX INFO: loaded from: classes7.dex */
@Metadata(d1 = {"\u00008\n\u0002\u0018\u0002\n\u0002\u0010\u0000\n\u0002\b\u0002\n\u0002\u0018\u0002\n\u0002\u0018\u0002\n\u0002\u0018\u0002\n\u0000\n\u0002\u0018\u0002\n\u0000\n\u0002\u0010\u0002\n\u0000\n\u0002\u0018\u0002\n\u0002\b\u0002\n\u0002\u0010 \n\u0002\u0010\b\n\u0000\bÇ\u0002\u0018\u00002\u00020\u0001B\u0007\b\u0002¢\u0006\u0002\u0010\u0002J\u000e\u0010\t\u001a\u00020\n2\u0006\u0010\u000b\u001a\u00020\fJ\u0014\u0010\r\u001a\u00020\u00052\f\u0010\u000e\u001a\b\u0012\u0004\u0012\u00020\u00100\u000fR\u001e\u0010\u0003\u001a\u0012\u0012\u0004\u0012\u00020\u00050\u0004j\b\u0012\u0004\u0012\u00020\u0005`\u0006X\u0082\u0004¢\u0006\u0002\n\u0000R\u000e\u0010\u0007\u001a\u00020\bX\u0082.¢\u0006\u0002\n\u0000¨\u0006\u0011"}, d2 = {"Lcom/zhufucdev/motion_emulator/data/MotionRecorder;", "", "()V", "callbacks", "Ljava/util/ArrayList;", "Lcom/zhufucdev/motion_emulator/data/MotionCallback;", "Lkotlin/collections/ArrayList;", "sensors", "Landroid/hardware/SensorManager;", "init", "", "context", "Landroid/content/Context;", "start", "sensorsRequired", "", "", "app_debug"}, k = 1, mv = {1, 9, 0}, xi = 48)
public final class MotionRecorder {
    private static SensorManager sensors;
    public static final MotionRecorder INSTANCE = new MotionRecorder();
    private static final ArrayList<MotionCallback> callbacks = new ArrayList<>();
    public static final int $stable = 8;

    private MotionRecorder() {
    }

    public final void init(Context context) {
        Intrinsics.checkNotNullParameter(context, "context");
        Object systemService = context.getSystemService((Class<Object>) SensorManager.class);
        Intrinsics.checkNotNullExpressionValue(systemService, "getSystemService(...)");
        sensors = (SensorManager) systemService;
    }

    /* JADX WARN: Type inference failed for: r0v3, types: [com.zhufucdev.motion_emulator.data.MotionRecorder$start$listener$1] */
    public final MotionCallback start(List<Integer> sensorsRequired) {
        Intrinsics.checkNotNullParameter(sensorsRequired, "sensorsRequired");
        final long start = System.currentTimeMillis();
        final ArrayList moments = new ArrayList();
        final Ref.ObjectRef callbackListener = new Ref.ObjectRef();
        final HashMap typedListeners = new HashMap();
        final List sensorsInvolved = CollectionsKt.toList(sensorsRequired);
        final int sensorCount = sensorsInvolved.size();
        final ?? r0 = new SensorEventListener() { // from class: com.zhufucdev.motion_emulator.data.MotionRecorder$start$listener$1
            /* JADX WARN: Multi-variable type inference failed */
            @Override // android.hardware.SensorEventListener
            public void onSensorChanged(SensorEvent event) {
                Function1<MotionMoment, Unit> function1;
                Function1<MotionMoment, Unit> function12;
                Intrinsics.checkNotNullParameter(event, "event");
                float elapsed = (System.currentTimeMillis() - start) / 1000.0f;
                if (!moments.isEmpty()) {
                    MotionMoment lastMoment = (MotionMoment) CollectionsKt.last((List) moments);
                    if (Math.abs(lastMoment.getElapsed() - elapsed) < 0.05f && !lastMoment.getData().containsKey(Integer.valueOf(event.sensor.getType()))) {
                        lastMoment.getData().put(Integer.valueOf(event.sensor.getType()), event.values.clone());
                        Function1<MotionMoment, Unit> function13 = typedListeners.get(Integer.valueOf(event.sensor.getType()));
                        if (function13 != null) {
                            function13.invoke(lastMoment);
                        }
                        if (lastMoment.getData().size() != sensorCount || (function12 = callbackListener.element) == null) {
                            return;
                        }
                        function12.invoke(lastMoment);
                        return;
                    }
                }
                Map data = new LinkedHashMap();
                data.put(Integer.valueOf(event.sensor.getType()), event.values.clone());
                MotionMoment moment = new MotionMoment(elapsed, data);
                moments.add(moment);
                Function1<MotionMoment, Unit> function14 = typedListeners.get(Integer.valueOf(event.sensor.getType()));
                if (function14 != null) {
                    function14.invoke(moment);
                }
                if (sensorCount != 1 || (function1 = callbackListener.element) == null) {
                    return;
                }
                function1.invoke(moment);
            }

            @Override // android.hardware.SensorEventListener
            public void onAccuracyChanged(Sensor sensor, int accuracy) {
            }
        };
        List<Integer> $this$forEach$iv = sensorsRequired;
        for (Object element$iv : $this$forEach$iv) {
            int it = ((Number) element$iv).intValue();
            SensorManager sensorManager = sensors;
            SensorManager sensorManager2 = null;
            if (sensorManager == null) {
                Intrinsics.throwUninitializedPropertyAccessException("sensors");
                sensorManager = null;
            }
            Sensor sensor = sensorManager.getDefaultSensor(it);
            SensorManager sensorManager3 = sensors;
            if (sensorManager3 == null) {
                Intrinsics.throwUninitializedPropertyAccessException("sensors");
            } else {
                sensorManager2 = sensorManager3;
            }
            sensorManager2.registerListener((SensorEventListener) r0, sensor, 2);
        }
        MotionCallback motionCallback = new MotionCallback() { // from class: com.zhufucdev.motion_emulator.data.MotionRecorder$start$result$1
            @Override // com.zhufucdev.motion_emulator.data.MotionCallback
            public Motion summarize() {
                SensorManager sensorManager4 = MotionRecorder.sensors;
                if (sensorManager4 == null) {
                    Intrinsics.throwUninitializedPropertyAccessException("sensors");
                    sensorManager4 = null;
                }
                sensorManager4.unregisterListener(r0);
                synchronized (MotionRecorder.INSTANCE) {
                    MotionRecorder.callbacks.remove(this);
                }
                String strRandomNanoId = NanoIdUtils.randomNanoId();
                Intrinsics.checkNotNullExpressionValue(strRandomNanoId, "randomNanoId(...)");
                return new Motion(strRandomNanoId, null, start, moments, sensorsInvolved);
            }

            /* JADX WARN: Multi-variable type inference failed */
            @Override // com.zhufucdev.motion_emulator.data.MotionCallback
            public void onUpdate(Function1<? super MotionMoment, Unit> l) {
                Intrinsics.checkNotNullParameter(l, "l");
                callbackListener.element = l;
            }

            @Override // com.zhufucdev.motion_emulator.data.MotionCallback
            public void onUpdate(int type, Function1<? super MotionMoment, Unit> l) {
                Intrinsics.checkNotNullParameter(l, "l");
                typedListeners.put(Integer.valueOf(type), l);
            }
        };
        synchronized (this) {
            callbacks.add(motionCallback);
        }
        return motionCallback;
    }
}

File: app/src/main/java/com/zhufucdev/motion_emulator/data/MotionRecorder.kt
package com.zhufucdev.motion_emulator.data

import android.content.Context
import android.hardware.Sensor
import android.hardware.SensorEvent
import android.hardware.SensorEventListener
import android.hardware.SensorManager
import com.aventrix.jnanoid.jnanoid.NanoIdUtils
import com.zhufucdev.me.stub.Motion
import com.zhufucdev.me.stub.MotionMoment

/**
 * 运动数据录制回调接口
 *
 * 用于接收传感器数据更新通知和获取录制结果
 */
interface MotionCallback {
    /**
     * 获取录制结果并停止录制
     *
     * 注销传感器监听器，汇总所有录制的传感器数据，
     * 生成 [Motion] 对象并返回
     *
     * @return 包含所有录制数据的 [Motion] 对象
     */
    fun summarize(): Motion

    /**
     * 注册通用数据更新监听器
     *
     * 当所有传感器类型都有数据时触发回调
     *
     * @param l 回调函数，接收 [MotionMoment] 数据
     */
    fun onUpdate(l: (MotionMoment) -> Unit)

    /**
     * 注册特定类型传感器数据更新监听器
     *
     * 仅当指定类型的传感器数据更新时触发回调
     *
     * @param type 传感器类型常量（如 [Sensor.TYPE_ACCELEROMETER]）
     * @param l 回调函数，接收 [MotionMoment] 数据
     */
    fun onUpdate(type: Int, l: (MotionMoment) -> Unit)
}

/**
 * 运动数据录制器（单例）
 *
 * 负责管理传感器监听和运动数据录制。支持同时录制多种传感器数据，
 * 并在传感器数据时间接近时合并为同一时刻的数据。
 *
 * ## 使用流程
 * 1. 调用 [init] 初始化传感器管理器
 * 2. 调用 [start] 开始录制，传入需要录制的传感器类型列表
 * 3. 通过返回的 [MotionCallback] 监听数据更新
 * 4. 调用 [MotionCallback.summarize] 获取录制结果
 *
 * ## 传感器合并策略
 * 当两个传感器事件的时间差小于 50ms 时，会将它们合并到同一个 [MotionMoment] 中，
 * 以模拟多传感器同时采样的效果。当所有传感器类型都有数据后，才会触发通用回调。
 */
object MotionRecorder {
    private lateinit var sensors: SensorManager
    private val callbacks = arrayListOf<MotionCallback>()

    /**
     * 初始化传感器管理器
     *
     * 必须在 [start] 之前调用
     *
     * @param context Android Context，用于获取系统服务
     */
    fun init(context: Context) {
        sensors = context.getSystemService(SensorManager::class.java)
    }

    /**
     * 开始录制传感器数据
     *
     * 为每个指定的传感器类型注册监听器，创建并返回 [MotionCallback] 实例
     *
     * @param sensorsRequired 需要录制的传感器类型列表（如 [Sensor.TYPE_ACCELEROMETER]）
     * @return [MotionCallback] 实例，用于监听数据更新和获取录制结果
     */
    fun start(sensorsRequired: List<Int>): MotionCallback {
        val start = System.currentTimeMillis()
        val moments = arrayListOf<MotionMoment>()
        var callbackListener: ((MotionMoment) -> Unit)? = null
        val typedListeners = hashMapOf<Int, (MotionMoment) -> Unit>()
        val sensorsInvolved = sensorsRequired.toList()
        val sensorCount = sensorsInvolved.size

        val listener = object : SensorEventListener {
            override fun onSensorChanged(event: SensorEvent) {
                val elapsed = (System.currentTimeMillis() - start) / 1000F

                // 传感器数据合并逻辑：
                // 如果最后一个 moment 的时间差小于 50ms，且该传感器类型尚未记录，
                // 则将数据合并到同一个 MotionMoment 中
                if (moments.isNotEmpty()) {
                    val lastMoment = moments.last()
                    if (Math.abs(lastMoment.elapsed - elapsed) < 0.05f
                        && !lastMoment.data.containsKey(event.sensor.type)
                    ) {
                        lastMoment.data[event.sensor.type] = event.values
                        typedListeners[event.sensor.type]?.invoke(lastMoment)
                        // 当所有传感器类型都有数据时，触发通用回调
                        if (lastMoment.data.size == sensorCount) {
                            callbackListener?.invoke(lastMoment)
                        }
                        return
                    }
                }

                // 创建新的 MotionMoment
                val data = mutableMapOf<Int, FloatArray>()
                data[event.sensor.type] = event.values.clone()
                val moment = MotionMoment(elapsed, data)
                moments.add(moment)
                typedListeners[event.sensor.type]?.invoke(moment)
            }

            override fun onAccuracyChanged(sensor: Sensor?, accuracy: Int) {}
        }

        sensorsRequired.forEach {
            val sensor = sensors.getDefaultSensor(it)
            sensors.registerListener(listener, sensor, SensorManager.SENSOR_DELAY_UI)
        }

        val result = object : MotionCallback {
            override fun summarize(): Motion {
                sensors.unregisterListener(listener)
                synchronized(MotionRecorder) {
                    callbacks.remove(this)
                }

                return Motion(NanoIdUtils.randomNanoId(), null, start, moments, sensorsInvolved)
            }

            override fun onUpdate(l: (MotionMoment) -> Unit) {
                callbackListener = l
            }

            override fun onUpdate(type: Int, l: (MotionMoment) -> Unit) {
                typedListeners[type] = l
            }
        }

        synchronized(this) {
            callbacks.add(result)
        }

        return result
    }
}
